File: 2337-move-pieces-to-obtain-a-string/2337-move-pieces-to-obtain-a-string.js
var canChange = function(start, target) {
    const n = start.length;
    let i = 0, j = 0;
    
    while (i < n || j < n) {
        while (i < n && start[i] === '_') {
            i++;
        }
        while (j < n && target[j] === '_') {
            j++;
        }
        
        if (i === n && j === n) {
            return true;
        }
        
        if (i === n || j === n || start[i] !== target[j]) {
            return false;
        }
        
        if (start[i] === 'L' && i < j) {  // L can only move left
            return false;
        }
        if (start[i] === 'R' && i > j) {  // R can only move right
            return false;
        }
        
        i++;
        j++;
    }
    
    return true;
};

File: 2337-move-pieces-to-obtain-a-string/2337-move-pieces-to-obtain-a-string.py
class Solution:
    def canChange(self, start: str, target: str) -> bool:
        n = len(start)
        i = j = 0
        
        while i < n or j < n:
            while i < n and start[i] == '_':  # Skip '_' in start
                i += 1
            while j < n and target[j] == '_':  # Skip '_' in target
                j += 1
            
            # Both strings are fully traversed
            if i == n and j == n:
                return True
            
            # One string finishes early
            if i == n or j == n or start[i] != target[j]:
                return False
            
            # Check constraints for L and R
            if start[i] == 'L' and i < j:  # L can only move left
                return False
            if start[i] == 'R' and i > j:  # R can only move right
                return False
            
            i += 1
            j += 1
        
        return True
